### PR TITLE
Add GTK support Close #797

### LINKED ISSUE #797: Rendering support for GTK platform
In an analogous way as it was done here (https://github.com/Redth/ZXing.Net.Mobile/pull/785) it would be great to have GTK support.

### PR DESCRIPTION
**Pull request by mierzynskim**
This PR will add rendering support for GTK. PR also includes example for both native GTK and XF backend.
Implementation of drawing an individual pixel may look like a little bit odd but I think that's the most readable version that can be achieved with GTK.
I wasn't sure how to tackle nuspec files for GTK but I assumed we can use the same convention as Xamarin Forms GTK backend and use `net45` as target framework (see [here](https://github.com/xamarin/Xamarin.Forms/blob/master/.nuspec/Xamarin.Forms.Platform.GTK.nuspec)).

### CODE REVIEW

**Comment by knocte** [comment]
hey @mierzynskim you targetted the wrong branch, can you apply to 'dev' branch instead?

File: Samples/GTK/gtk-gui/MainWindow.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		if ((this.Child != null))
+		{
+			this.Child.ShowAll();
+		}
+		this.DefaultWidth = 400;
+		this.DefaultHeight = 300;
+		this.Show();
+		this.DeleteEvent += new global::Gtk.DeleteEventHandler(this.OnDeleteEvent);
+	}
+}
```
Comment (knocte) [inline Samples/GTK/gtk-gui/MainWindow.cs]: nit: use space instead of tabs in this file

File: Samples/GTK/gtk-gui/generated.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+		{
+			return Stetic.ActionGroups.GetActionGroup(type.FullName);
+		}
+
+		public static Gtk.ActionGroup GetActionGroup(string name)
+		{
+			return null;
+		}
+	}
+}
```
Comment (knocte) [inline Samples/GTK/gtk-gui/generated.cs]: same nit here: spaces instead of tabs

**Comment by knocte** [comment]
Mmm, CI doesn't pass because of a weird NuGet error related to .sln or .csproj files:

```
Executing: "C:/projects/zxing-net-mobile/tools/NuGet.exe" restore "C:/projects/zxing-net-mobile/ZXing.Net.Mobile.Forms.sln" -NonInteractive
MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\bin'.
Error parsing solution file at C:\projects\zxing-net-mobile\ZXing.Net.Mobile.Forms.sln: Exception has been thrown by the target of an invocation.  Error parsing the nested project section in solution file. A project with the GUID "{A7E26AE0-12AF-4CFD-9B1E-3101347CC6F4}" is listed as being nested under project "{2DC0C8D5-DF7F-4D96-886D-D7E83522C9BF}", but does not exist in the solution.  C:\projects\zxing-net-mobile\ZXing.Net.Mobile.Forms.sln
An error occurred when executing task 'libs'.
Error: System.AggregateException: One or more errors occurred. ---> Cake.Core.CakeException: NuGet: Process returned an error (exit code 1).
```

Did you edit these files manually instead of letting VS change them? :)

File: Samples/GTK/gtk-gui/MainWindow.cs (hunk 1 above)
Comment (mierzynskim) [reply Samples/GTK/gtk-gui/MainWindow.cs]: I think we can change it because it's autogenerated file 😞

File: Samples/GTK/gtk-gui/generated.cs (hunk 2 above)
Comment (mierzynskim) [reply Samples/GTK/gtk-gui/generated.cs]: same as above

File: Samples/GTK/gtk-gui/MainWindow.cs (hunk 1 above)
Comment (knocte) [reply Samples/GTK/gtk-gui/MainWindow.cs]: mkay

**Comment by mierzynskim** [comment]
Hmm, I think CI doesn't have GTK installed... I thought it's shipped with Mono by default. I'll add it as external library

**Comment by knocte** [comment]
Not sure adding it as external lib is the best approach. Maybe download the .msi file that installs these libs and install it non-interactively from the cake files?

**Comment by knocte** [comment]
Or this? https://chocolatey.org/packages/gtksharp

**Comment by mierzynskim** [comment]
Hard to say. That's the approach XF chose https://github.com/xamarin/Xamarin.Forms/tree/master/Xamarin.Forms.Platform.GTK/Libs

**Comment by knocte** [comment]
ah ok! if that's the case, you could do the same I guess, just grab the exact same binaries

File: Source/ZXing.Net.Mobile.Forms.WindowsUniversal/ZXing.Net.Mobile.Forms.WindowsUniversal.csproj [hunk 3]
```diff
@@ -131,7 +131,9 @@
     </ProjectReference>
   </ItemGroup>
   <ItemGroup>
-    <Reference Include="netstandard"/>
+    <Reference Include="netstandard">
+      <HintPath>Libs\netstandard.dll</HintPath>
```
Comment (knocte) [inline Source/ZXing.Net.Mobile.Forms.WindowsUniversal/ZXing.Net.Mobile.Forms.WindowsUniversal.csproj]: adding this binary lib is a bit ugly (and doesn't seem to work anyway), how about you revert this commit and try to add the nuget package called "NETStandard.Library"

Comment (mierzynskim) [reply Source/ZXing.Net.Mobile.Forms.WindowsUniversal/ZXing.Net.Mobile.Forms.WindowsUniversal.csproj]: Unfortunately, it doesn't work with that nuget package 😢. It gives the same error as before. That's a nasty workaround but it works. The best solution would be to update target project to [newer ones](https://stackoverflow.com/questions/51211144/the-type-attribute-is-defined-in-an-assembly-that-is-not-referenced-xamarin) but I think that's a task for another PR.

**Comment by knocte** [comment]
@redth FYI I tested this and it works perfectly, can you review/merge? thanks!

**Comment by mierzynskim** [comment]
@Redth I think we have a final version now. I also tested it and it looks like it works fine. Please have a look when you have some time.

**Comment by rblanca** [comment]
Any news? Thanks

**Comment by knocte** [comment]
@rblanca we have this merged in this fork in case you're interested: https://github.com/nblockchain/ZXing.Net.Xamarin (we also publish this fork on nuget)

**Comment by knocte** [comment]
@Redth can you include this in 3.0?

**Comment by Redth** [comment]
@knocte I'm super hesitant to add this into the main package, mostly because https://github.com/Redth/ZXing.Net.Mobile/pull/798/files?file-filters%5B%5D=.nuspec#diff-978f915984ac9573cd5e42dfc0550c3fR82-R83

The TFM of `net45` isn't really appropriate to use in the main package if the binaries are specifically for GTK apps.  

What we maybe _could_ do is release a separate package of `ZXing.Net.Mobile.GTK` with these binaries in it.

**Comment by Redth** [comment]
@knocte any interest in trying to rebase this PR's work on current master?

**Comment by knocte** [comment]
> The TFM of net45 isn't really appropriate to use in the main package if the binaries are specifically for GTK apps.

What do you mean? what's the issue exactly? I've been using this code for a while now and works perfectly. The proof is this fork we publish on nuget: https://github.com/nblockchain/ZXing.Net.Xamarin and the app that consumes it: https://github.com/nblockchain/geewallet

**Comment by Redth** [comment]
The problem is if I ship code that depends on GTK into an assembly which is in the net45 TFM, and someone tries to install and use the package in say a windows console app, they aren’t going to have a good time of it. Because there is no TFM that exists to pivot on GTK support directly, there’s limited options. 

1. Ship it in a separate nuget package such as `ZXing.Net.Mobile.GTK` (and use net45 for that) - this is a bit more explicit since you have to opt in by installing the nuget package intentionally which means you likely understand what you are doing and won’t try and use it in the wrong type of app. 

2. See if there’s any sort of MSBuild property that is exposed when an app uses GTK and use a .targets file in the main nuget package to pivot on which assembly gets referenced. This is kind of weird if there isn’t any other non-GTK implementation shipping (which there currently is not). 

Basically 1 is less discoverable but much easier to implement than 2.

I’m leaning towards option 1.

**Comment by knocte** [comment]
> and use the package in say a windows console app, 

But this is not supported by ZXing.Net.Mobile anyway AFAIU, why not just put a warning about this on the Readme?

**Comment by Redth** [comment]
It's not supported _today_, but it _could_ be one day.  That alone may not be a good enough reason, but generally the practice we use at Microsoft is to use a separate package name for GTK stuff.  Take `Xamarin.Forms.Platform.GTK` for instance.  

I think the right thing to do here is the separate package (and one for the Forms implementation too).  This means separate .csproj's which I don't love, but I think this is the right choice

**Comment by knocte** [comment]
> It's not supported today, but it could be one day.

One day? AFAIU net45 is the old framework; and these days in the .NET community people are even advocating to stop supporting it because "legacy" and just support .NETCore, why would anyone go in the other direction?

Anyway I understand the point of view (I use GTK myself with Xamarin.Forms) but it looks a bit ugly (because the other nonGTK platforms don't have a separate nuget package).

**Comment by Redth** [comment]
Yes, net45 is old, however net5 is coming, the same problem will exist there.  There will be no platform TFM for GTK pivoting.  The point is more about unintentionally expressing support for a TFM which you do not _actually_ support.  Yes this is still true of a `.GTK` package but as I said, it's a bit more explicit by naming and the intentional additional install helps to avoid confusion.

Yeah more nuget refs in a project stinks, but I definitely think the trade off is ok in this case.

**Comment by knocte** [comment]
> any interest in trying to rebase this PR's work on current master?

I have interest :) What I'm not sure is if I have time. Especially considering that even if I did the effort of rebasing it, I couldn't adopt the new 3.x verison of ZXing.Net.Mobile right away, because I just realised that you introduced Xamarin.Essentials as a dependency :( What was the motivation for that? It basically would break our setup because XamEssentials doesn't support macOS and Linux+GTK platforms.
